### DISCUSSION TITLE
Distributing static assets

### DISCUSSION THREAD

**Discussion by radu-matei** [bytecodealliance/SIG-Registries#9] [General]
Static assets are an area we have explored in the past, and being able to distribute static assets is part of the feature list for the registry MVP (see I and J in #3)

There are two approaches to this that we have discussed:

- packaging static assets in custom data sections as part of the Wasm component itself
- distributing them separately

There are certain advantages and disadvantages in both scenarios.
(I will do a more complete write-up in the coming week(s), but I wanted to start this discussion thread so we can capture people's opinions in a more structured way).

cc @nitsky

**Comment by lann** [comment]
Maybe you intend for "distribute them separately" to cover this, but "sub-options" of that would include packaging them together in some e.g. ZIP file, or actually splitting them into separate e.g. URLs to download.

**Reply by radu-matei** [reply]
You are correct.
By "separately" I only meant "not in the Wasm component", good point that there are lots possibilities there.
(I would expect being able to de-duplicate at upload, so a ZIP file doesn't seem like it would fit.)

**Reply by lann** [reply]
@kulakowski-wasm Would it be at all sensible to investigate Fuchsia's [FAR](https://fuchsia.dev/fuchsia-src/development/source_code/archive_format) for this? ZIP's internal compression and TAR's lack of index make them pretty unsuitable for this, but FAR doesn't seem to have either of those problems.

cc @fibonacci1729

**Comment by nitsky** [comment]
Including static assets in the component itself will mean that any change in the code will require re-downloading all the assets, and vice versa. Separating them will result in better caching. However, if they are distributed separately, the host will need to have runtime support for "connecting" a running component with the static assets it depends upon.

**Reply by kulakowski-wasm** [reply]
Context about Fuchsia and me for those who may not know: I worked on Fuchsia for a long, long time :) kernel, ipc, driver model, libc-ish things, other low level stuff, and was very adjacent to the package/component model aspects. And a couple bits of disclaimer! About my knowledge of it: it was a bit ago, things change, I know maybe more about design intentions than what shipped or will ship, my memory's fallible and full of just so stories, etc. There's tons of design constraints that Fuchsia's storage/blob/package/component models, and our Registry and Component models will share, and also lots of differences. Not to mention that everything has these really annoying path dependencies that prevent you from shipping the actual thing you dearly wanted to :)

Lann, I think it does make sense to investigate FAR! I'm curious how you find the documentation. I probably have both outdated assumptions/context, as well as actually useful implicit knowledge. I'm still excited to look / prepare feedback about it, but am very sure that other perspectives on it are essential too :)

So about downloading, nitsky: To Lann's point, Fuchsia does address that. The package filesystem is layered on top of a blob filesystem which is able to do fancy rolling deduplication schemes. And by "fancy" I mean that an algorithm can discover overlapping contents with high probability, even if those contents are not named or "aligned" in some way (like via a merkle tree, or just an ordinary filesystem hierarchy). And (maybe, this is one I don't remember) you can send diffs of blobs over the wire in either prod or dev workflows.

These combine to address use cases like fetching/uploading an object with substantially similar contents like big binary assests, and deduplicating storage across packages on a device. This is a reason to still use shared libraries inside a Fuchsia package, even though from the perspective of linking inside a single package you could easily imagine only static linking or just LTOing everything together, right, it's hermetic so you can see all of it at once kinda definitionally. (It's of course not the only reason to use dynamic linking.)

It's not immediately clear to me how coupled these (sending diffs; depduing on disk) are, and how useful they are to different entities and use cases.

For example, on-disk object deduplication may be highly valuable for relatively persistent storage of a set of objects curated by a single party, and stored on the smallest cheapest NAND you can find. Object deduplication may be less valuable to a more open ended and heterogenous object store backed by a hyperscaler's storage system.

Spending server cpu time to compute diffs that get sent over LTE may be worthwhile. But for a local development cycle usecase, "sending" a new object may be as cheap as changing a hardlink or whatever.

Running a blaze-like virtual filesystem for blob storage may be highly compelling to some and anathema to others.

Write amplification induced by one storage layer onto a lower layer may be terrifying or may not even occur to you.

You get the idea.

This message is already too long, but I want to raise one meta point. We're going to define APIs and write some code implementing various ends of those APIs. Those will not be the only ways people interact with registries! Different implementations and use cases are going to have other mechanisms of interaction that are outside of the standards and the reference implementations we define. This is a normal thing and a good thing.

It's important for us to allow ourselves to acknowledge the existence of some use cases, or some costs, or some complexity, and not entirely solve them. That doesn't mean that others (including the exactly us in our non-standards-bodies roles!) won't solve them, or that we won't offer affordances towards solving them.

**Reply by carlsverre** [reply]
Bit off topic, but reading through the docs on blobfs, I don't see where it explains the fancy rolling deduplication algorithm you mentioned. It appears that blobfs uses a merkle tree to verify each file, which should restrict it to deduplicating in an aligned way. Here is a link to the docs I found: https://fuchsia.dev/fuchsia-src/concepts/filesystems/blobfs

**Reply by kulakowski-wasm** [reply]
It may not have been implemented then! It does look like random access aware compression was, though. I wonder if there are opposing constraints there?

Anyway, https://arxiv.org/pdf/2104.02158.pdf is representative of research in this area I think.

And my point in bringing it up was just to point out that there may be more options than listed originally, not necessarily to advocate for it :)

**Reply by lann** [reply]
To clarify, my suggestion to explore FAR was purely about the file format, which has some nice properties for us. Fuchsia does a lot of cool stuff, but I didn't intend to dump the _whole_ can of worms out.

**Reply by kulakowski-wasm** [reply]
Oh absolutely. Yeah, my comment replied both to you, indirectly, and to nitsky's comment, and rambling about different but related fuchsia tech was relevant to both.

Thinking back to why I typed all I did: I think I found "any change in the code will require re-downloading all the assets, and vice versa" to be too absolute. Minimizing bytes sent over the wire vs filesystem performace vs filesystem capacity vs complexity, are core to what other packaging and updating and distribution systems deal with. They make different tradeoffs. The design space feels big still!

**Comment by lukewagner** [comment]
I think we should aim for supporting both options (single file and separate files) since both have strong use cases.  To @nitsky's comment above, I think the key is how to specify the runtime support when they are separate files -- if we're not careful, we'll add some ad hoc file system mechanism that will lead to all sorts of problems.  The approach I'm in favor of frames single-file and multi-file as different *representations* of the same logical component (so there would be 3 representations of a component: `.wat`, `.wasm` and "directory"), so that the runtime behavior is fully defined by the component model spec without anything else.

In the single-file scenario, the idea is that the component `.wasm` embeds static assets as [data segments](https://webassembly.github.io/spec/core/binary/modules.html#binary-data), which can then be used to implement whatever higher-level (libc) filesystem APIs in terms of the existing [`memory.init`](https://webassembly.github.io/spec/core/syntax/instructions.html#syntax-instr-memory) instructions.

Then, for the multi-file scenario, we can say that a single component gets expanded into a directory where:
* Each data segment turns into a file in the directory.  If there is a [name section](https://webassembly.github.io/spec/core/appendix/custom.html#name-section) present which assigns a name to the data segment, that name is used.  Otherwise, the name is the data segment index (the same index used as an immediate in `memory.init` instructions).
* Each core module nested in the component turns into a separate `.wasm` file (using the same name section / index naming scheme as `data` segments).
* Each component nested in the component turns into a nested directory (recursively).
* The original component replaces the original nested contents with an `import` that uses a [relative path](https://datatracker.ietf.org/doc/html/rfc3986#section-4.2) as the import string so that normal relative path resolution resolves the import to same definition that was originally nested.  This remaining "husk" serves only to link everything together (via `import` and `instance` definitions), taking advantage of the fact that components are essentially [executable linker scripts](https://docs.google.com/presentation/d/11lY9GBghZJ5nCFrf4MKWVrecQude0xy_buE--tnO9kQ/edit#slide=id.gf5785162e3_0_3).

So, for example, the single-file component:
```wasm
;; my.wasm
(component
  (module $M (@name "guts.wasm")
    ... wasm code
    (data $index_html (@name "index.html") "<!doctype html> ...")
  )
  (instance $m (instantiate $M))
  ...
)
```
would be transformed into a directory:
* `my/`
  * `my.wasm`
  * `guts.wasm`
  * `index.html`

containing:
```wasm
;; my.wasm
(component
  (import "./guts.wasm" (module $M ...module type...))
  (instance $m (instantiate $M))
  ...
)
```
```wasm
;; guts.wasm
(module
  (import "./index.html" (data $index_html))
  ... wasm code
)
```
Importantly, there's no runtime WASI filesystem dependency added here and thus no subtle filesystem questions (is it mutable? persistent? shared and how widely and what about name collisions?).

The one thing we need, spec-wise, to make this work is supporting `data` imports in core wasm.  Because of the "M" in MVP, `data` imports/exports were conspicuously left out of the WebAssembly 1.0 but they have been discussed and wished-for a number of times, so I was thinking I'd champion this core wasm proposal.  In the short term, I think it should be pretty easy to implement as part of the native component support in wasmtime.  (It's also easy enough to polyfill onto 1.0 by wrapping the data file in a trivial core wasm module that exports a function wrapping `memory.init`.)

**Reply by kulakowski-wasm** [reply]
I like the separation into a logical representation and specific spec'd physical instantiations of that.

It also seems like `data` imports need to and probably will happen eventually, regardless of the details here?

I can't tell how general you intend "directory" to be by putting it in quotes like that. Is it the like least common denominator of FAT and APFS and NTFS and .... ? Is it intended to be anything that can serve up a hierarchical namespace?

Do I have to stick directory OR data segments? Can I have both a data segment, and a parallel file?

**Reply by lukewagner** [reply]
Yes, for `data` imports it's a "when" not "if" question, I believe.

"Directory" would I suppose would cover any system that can interpret relative URIs, so it could include both file systems and HTTP.  Naming is definitely open for discussion, though.

That's a good question of whether it's all-or-nothing; I suppose nothing *forces* all the data segments to be out-of-line -- if there was a `wasm2dir` tool, you could imagine it has a default behavior of expanding everything, but there could be options for saying what you did/didn't want out-of-line.  The reverse `dir2wasm` would be happy with whatever.

**Reply by lann** [reply]
re: "Directory", one of the things that appeals to me about the FAR format is its [restrictions on paths](https://fuchsia.dev/fuchsia-src/development/source_code/archive_format#path_data), which basically boil down to "a flattened subtree with no relative path funny business".

**Reply by lann** [reply]
For files embedded in data segments, would that look to the registry like a single blob or like multiple blobs in a sort of "Wasm archive"? If "single blob", things like deduplication or delta-synced-updates become complicated. If "multiple blob", signature verification is potentially complicated.

**Reply by lukewagner** [reply]
I think the registry would always use the directory (multiple blob) representation, so that each individual file had its own signature and de-duping followed naturally.  Then yes, the directory as a whole would need a signature (merkle tree style) and you're right that it's complicated in that we have to precisely define the input data to the hash function... but I think that's what we have to do in any case with these merkle tree schemes?

**Reply by lukewagner** [reply]
Oh, and regarding paths, I think we should definitely restrict the relative paths in this "directory representation" to a very well-behaved, maximally-portable subset of relative URIs.

**Reply by lann** [reply]
re: multi-file signatures, I'm more worried about how to precisely define the Wasm binary's hash if it is embedding other files as well. Maybe it's relatively straightforward like "just skip the files", but it's the kind of thing that can be hard to spec in a way that doesn't invite vulnerable implementations.

**Reply by lukewagner** [reply]
I think you're right to be wary about this in general, but what's neat about the scheme described above is that, in the multi-file directory representation, we have already explicitly removed the nested files in `my/my.wasm` (replacing them by `import`s), so `my/my.wasm` can just be a simple blob -- the only thing that's "interesting" is how to sign the directory `my/` containing the 3 blobs.

**Reply by lann** [reply]
This latest version of the Publish API proposal includes a mechanism for signing a ~~tree~~ list (with paths, implicitly forming a tree) of content: https://github.com/bytecodealliance/SIG-Registries/blob/proposal/publish-api/proposals/Publish-API.md#content-digest

**Reply by lukewagner** [reply]
Cool!  So, then, could the Publish API give a component-package sub-paths for the individual blobs of the directory representation of component so that you could *either* POST/GET a single-file `.wasm` to the package's root path (e.g., `GET /components/bytecodealliance.org:dog-facts/v1.0.0` *or* POST/GET the individual blobs of the directory representation via the subpaths (e.g., `GET /components/bytecodealliance.org:dog-facts/v1.0.0/dog-database.json`)?

**Reply by lann** [reply]
I think so! My goal is to support fetching from a variety of storage layouts including single archive files. If your data section proposal defines a precise "lossless" transformation to/from a "directory" of multiple files then it should be possible to construct an index from that format.
